tuning/policydrivencontroller/pdt_callback.py

```diff
@@ -0,0 +1,236 @@
+    """Implements the policy driven trainer loop control based on policy definition file and metrics"""
+    
+    def __init__(self, train_control_args, training_args):
+        """Initializes the callback for policy-driven trainer control.
+
+        Args:
+            train_control_args: File path for trainer control definition file
+            training_args: TrainingArguments object
+        """
+        self.__controllers = {}
+        if os.path.exists(train_control_args.training_control_definition_file):
+            with open(train_control_args.training_control_definition_file, "r") as f:
+                self.training_control_def = yaml.safe_load(f)
+                for controller in self.training_control_def['controllers']:
+                    name = controller['name']
+                    controller_metric_objs = []
+                    for cm in controller['controller-metrics']:
+                        obj = None
+                        try:
+                            # Get the controller-metric class type
+                            class_type = getattr(contmetrics, cm['handler-class'])
+                            # Initialize the controller-metric instance
+                            if 'arguments' in cm:
+                                obj = class_type(cm['name'], cm['arguments'])
+                            else:
+                                obj = class_type(cm['name'])
+                        except Exception as e:
+                            logger.fatal(e)
+                        assert (obj.validate(training_args)), 'Controller metric class [%s] cannot be computed because the training args do not support it' % (cm['handler-class'])
+                        controller_metric_objs.append(obj)
+                    self.__controllers[name] = controller_metric_objs
+        else:
+            raise ValueError("Controller configuration [%s] does NOT exist" % train_control_args.training_control_definition_file)
+
+    def __apply_control(self, cb, control):
+        """Given a controller-block, applies the control operation to the training loop.
+
+        Args:
+            cb: Controller block dictionary
+            control: TrainerControl object
+
+        Returns:
+            None.
+        """
+        if 'should_training_stop' in cb['control-operation']:
+            control.should_training_stop = cb['control-operation']['should_training_stop']
+        elif 'should_epoch_stop' in cb['control-operation']:
+            control.should_epoch_stop = cb['control-operation']['should_epoch_stop']
+        elif 'should_save' in cb['control-operation']:
+            control.should_save = cb['control-operation']['should_save']
+        elif 'should_evaluate' in cb['control-operation']:
+            control.should_evaluate = cb['control-operation']['should_evaluate']
+        elif 'should_log' in cb['control-operation']:
+            control.should_log = cb['control-operation']['should_log']
```

Comment:
can we use a for loop through the operations listed in `control-operation`, so that if it changes, it gets automatically supported?

Reply:
done